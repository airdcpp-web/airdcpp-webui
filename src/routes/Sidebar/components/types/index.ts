//import * as React from 'react';

import { IconType } from 'components/semantic/Icon';

import * as API from 'types/api';
import * as UI from 'types/ui';

export interface SessionLayoutLabels {
  // Label for button that opens a new session
  newCaption?: React.ReactNode;

  // Label for button that opens a new session
  newDescription?: React.ReactNode;

  // Label for button that opens a new session
  newIcon?: IconType;
}

export interface SessionLayoutManageProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> {
  // Unique ID of the section (used for storing and loading the previously open tab)
  baseUrl: string;

  // Array of the items to list
  items: SessionT[];

  // Session actions to show in the action menu
  actionIds?: string[];

  uiActions: UI.ModuleActions<SessionT, void, UIActionT>;
  sessionApi: UI.SessionActions<SessionT> & SessionApiT;

  // Item ID that is currently active (if any)
  activeId: API.IdType | undefined;

  // AccessConstant defining whether the user has edit permission
  editAccess: API.AccessEnum;

  unreadInfoStore: UI.UnreadInfoStore;

  remoteMenuId: string;
}

interface SessionMainLayoutComponents {
  newButton: React.ReactElement<any> | null;
  listActionMenuGetter: () => React.ReactNode;
  itemHeaderTitle: React.ReactNode;
  itemHeaderDescription: React.ReactNode;
  itemHeaderIcon: IconType;
  sessionMenuItems: React.ReactNode[];
}

export interface SessionMainLayoutProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionsT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> extends Pick<
      SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>,
      'unreadInfoStore'
    >,
    SessionMainLayoutComponents {
  activeItem: SessionT | null;
  onKeyDown: (event: React.KeyboardEvent) => void;
  children: React.ReactNode;

  actions: SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>['uiActions'];
}

export type SessionChildProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionsT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> = Pick<
  SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>,
  'sessionApi' | 'uiActions'
> &
  UI.RouteComponentProps & {
    session: SessionT;
    sessionT: UI.ModuleTranslator;
  };

export interface NewSessionLayoutProps extends UI.RouteComponentProps {
  sessionT: UI.ModuleTranslator;
}

export interface SessionLocationState {
  pending?: boolean;
}
