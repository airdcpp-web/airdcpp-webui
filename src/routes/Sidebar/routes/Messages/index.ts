export { default } from './components/Messages';
