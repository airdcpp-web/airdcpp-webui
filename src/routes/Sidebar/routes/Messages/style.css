.ccpm-state {
  display: flex;
}

.ccpm-state > i {
  align-self: center;
}
