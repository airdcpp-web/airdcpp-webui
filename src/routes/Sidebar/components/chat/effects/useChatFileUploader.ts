import * as React from 'react';

import NotificationActions from 'actions/NotificationActions';

import * as UI from 'types/ui';

import { toI18nKey } from 'utils/TranslationUtils';
import { useFormatter } from 'utils/ValueFormat';

interface UseFileUploaderProps {
  appendText: (text: string) => void;
  handleFileUpload: UI.ChatFileUploadHandler;
  t: UI.TranslateF;
}

export const useFileUploader = ({
  appendText,
  handleFileUpload,
  t,
}: UseFileUploaderProps) => {
  const { formatSize } = useFormatter();
  const [files, setFiles] = React.useState<File[] | null>(null);
  const [uploading, setUploading] = React.useState(false);

  const resetFiles = () => {
    setFiles(null);
  };

  const onDropFile = (acceptedFiles: File[]) => {
    const maxSize = 100 * 1024 * 1024;

    const files = acceptedFiles.filter((file) => {
      if (file.size > maxSize) {
        NotificationActions.error({
          title: file.name,
          message: t(toI18nKey('fileTooLarge', UI.Modules.COMMON), {
            defaultValue: 'File is too large (maximum size is {{maxSize}})',
            replace: {
              maxSize: formatSize(maxSize),
            },
          }),
        });

        return false;
      }

      return true;
    });

    if (!files.length) {
      return;
    }

    setFiles(files);
  };

  const onUploadFiles = async (files: File[]) => {
    resetFiles();
    setUploading(true);

    for (const file of files) {
      try {
        const res = await handleFileUpload(file);
        appendText(res.magnet);
      } catch (e) {
        NotificationActions.apiError('Failed to upload the file', e);
      }
    }

    setUploading(false);
  };

  const onPaste = (evt: React.ClipboardEvent<HTMLTextAreaElement>) => {
    if (
      evt.clipboardData &&
      evt.clipboardData.files &&
      (evt.clipboardData as any).files.length
    ) {
      const files: File[] = [];

      {
        // DataTransferItemList isn't a normal array, convert it first
        // https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList
        const dataTransferItemList = (evt.clipboardData as any).files as any[];
        for (let i = 0; i < dataTransferItemList.length; i++) {
          files.push(dataTransferItemList[i]);
        }
      }

      evt.preventDefault();
      onDropFile(files);
    }
  };

  return {
    files,
    uploading,
    onDropFile,
    onPaste,
    onUploadFiles,
    resetFiles,
  };
};
