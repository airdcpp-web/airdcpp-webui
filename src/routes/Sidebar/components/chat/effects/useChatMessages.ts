import { useEffect, useState } from 'react';

import * as API from 'types/api';
import * as UI from 'types/ui';

/*export interface ChatSession extends UI.SessionItemBase {
  hub_url?: string;
}

export interface ChatLayoutProps extends UI.ChatController {
  chatAccess: string;
  messageStore: UI.SessionMessageStore;
  highlightRemoteMenuId: string;
}*/

export const useChatMessages = (
  session: UI.SessionItemBase,
  messageStore: UI.SessionMessageStore,
  chatAPI: UI.ChatAPI
) => {
  const [messages, setMessages] = useState<UI.MessageListItem[] | null>([]);

  useEffect(() => {
    // Session changed, update the messages
    if (!messageStore.isSessionInitialized(session.id)) {
      setMessages(null);
      chatAPI.fetchMessages(session);
    } else {
      const sessionMessages = messageStore.getSessionMessages(session.id);
      if (sessionMessages) {
        setMessages(sessionMessages);
      }
    }
  }, [session.id]);

  useEffect(() => {
    // Subscribe for new messages
    const unsubscribe = (messageStore as any).listen(
      (newMessages: UI.MessageListItem[], id: API.IdType) => {
        if (id !== session.id) {
          return;
        }

        setMessages(newMessages);
      }
    );

    return unsubscribe;
  }, [session.id]);

  return messages;
};
