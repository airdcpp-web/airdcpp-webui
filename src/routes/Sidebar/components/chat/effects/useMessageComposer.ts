import React, { useEffect } from 'react';

import { OnChangeHandlerFunc } from 'react-mentions';
import { useLocation } from 'react-router';
import * as UI from 'types/ui';

import {
  loadSessionProperty,
  removeSessionProperty,
  saveSessionProperty,
} from 'utils/BrowserUtils';
import ChatCommandHandler from '../ChatCommandHandler';

import { Location, useNavigate } from 'react-router';
import { useSocket } from 'context/SocketContext';
import { useSession } from 'context/SessionContext';

const getStorageKey = (location: Location) => {
  return `last_message_${location.pathname}`;
};

const loadState = (location: Location) => {
  const text = loadSessionProperty(getStorageKey(location), '');
  return text;
};

const saveState = (text: string, location: Location) => {
  if (text) {
    saveSessionProperty(getStorageKey(location), text);
  } else {
    removeSessionProperty(getStorageKey(location));
  }
};

interface MessageComposerProps {
  t: UI.TranslateF;
  chatController: UI.ChatController;
}

export const useMessageComposer = ({ chatController, t }: MessageComposerProps) => {
  const session = useSession();
  const location = useLocation();
  const navigate = useNavigate();
  const [inputText, setInputText] = React.useState('');
  const socket = useSocket();

  const handleCommand = (commandText: string) => {
    let command, params;

    {
      // Parse the command
      const whitespace = commandText.indexOf(' ');
      if (whitespace === -1) {
        command = commandText.substring(1);
      } else {
        command = commandText.substring(1, whitespace);
        params = commandText.substring(whitespace + 1);
      }
    }

    ChatCommandHandler(chatController).handle(command, params, {
      location,
      navigate,
      t,
      socket,
      session,
    });
  };

  const updateText = (newText: string) => {
    saveState(newText, location);
    setInputText(newText);
  };

  const appendText = (toAppend: string) => {
    let newText = inputText;
    if (newText) {
      newText += ' ';
    }
    newText += toAppend;

    updateText(newText);
  };

  const handleSend = (textToSend: string) => {
    const { chatApi, session } = chatController;
    chatApi.sendChatMessage(session, textToSend);
  };

  useEffect(() => {
    setInputText(loadState(location));
  }, [location.pathname]);

  const onTextChanged: OnChangeHandlerFunc = (event, markupValue, plainValue) => {
    updateText(plainValue);
  };

  const sendText = () => {
    // Trim only from end to allow chat messages such as " +help" to be
    // sent to other users
    // This will also prevent sending empty messages
    const textTrimmed = inputText.replace(/\s+$/, '');

    if (textTrimmed) {
      if (textTrimmed[0] === '/') {
        handleCommand(textTrimmed);
      }

      handleSend(textTrimmed);
    }

    updateText('');
  };

  const onKeyDown = (event: React.KeyboardEvent) => {
    if (event.key === 'Enter' && !event.shiftKey) {
      event.preventDefault();
      sendText();
    }
  };

  return {
    sendText,
    text: inputText,
    onKeyDown,
    onTextChanged,
    appendText,
  };
};
