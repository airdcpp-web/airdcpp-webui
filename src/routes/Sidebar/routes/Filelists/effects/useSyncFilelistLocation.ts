import * as React from 'react';

import * as API from '@/types/api';
import * as UI from '@/types/ui';

import { changeFilelistDirectory } from '@/services/api/FilelistApi';
import { runBackgroundSocketAction } from '@/utils/ActionUtils';
import { Location, useNavigate } from 'react-router';

export type FilelistLocationState = { directory: string } | undefined;

interface ListBrowserProps {
  session: API.FilelistSession;
  location: Location;
  sessionT: UI.ModuleTranslator;
}

// Keeps the current location in sync with the backend
export const useSyncFilelistLocation = ({
  session,
  location,
  sessionT,
}: ListBrowserProps) => {
  const navigate = useNavigate();

  const apiSendChangeDirectory = (directory: string) => {
    runBackgroundSocketAction(
      () => changeFilelistDirectory(session, directory),
      sessionT.plainT,
    );
  };

  const updateLocationState = (directory: string, replace: boolean) => {
    const locationState = location.state as FilelistLocationState;
    if (directory === locationState?.directory) {
      return;
    }

    // Handle it through location state data
    // as we also handle history navigation changes
    const newState: FilelistLocationState = {
      directory,
    };

    navigate(location.pathname, {
      state: newState,
      replace,
    });
  };

  // Handler for directory changes received from the API
  React.useEffect(() => {
    const locationState = location.state as FilelistLocationState;

    if (!locationState) {
      // Initial mount
      updateLocationState(session.location!.path, true);
      return;
    }

    // Nothing has changed?
    if (locationState.directory === session.location!.path) {
      return;
    }

    // Change initiated by another session/GUI, update our location
    updateLocationState(session.location!.path, true);
  }, [session.location?.path]);

  // Handler for own changes
  React.useEffect(() => {
    const locationState = location.state as FilelistLocationState;
    if (!locationState) {
      // Initial mount, handled by the previous useEffect
      return;
    }

    const directory = locationState.directory;
    if (directory === session.location!.path) {
      return;
    }

    // It's our change
    if (directory) {
      apiSendChangeDirectory(directory);
    }
  }, [(location.state as FilelistLocationState)?.directory]);

  return updateLocationState;
};
