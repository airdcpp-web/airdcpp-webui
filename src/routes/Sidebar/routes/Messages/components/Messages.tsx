import * as React from 'react';

import SessionLayout from 'routes/Sidebar/components/SessionLayout';
import UserItemHandlerDecorator from 'routes/Sidebar/decorators/UserItemHandlerDecorator';

import PrivateChatSessionStore from 'stores/PrivateChatSessionStore';
import PrivateChatAPIActions from 'actions/reflux/PrivateChatActions';

import MessageNew from 'routes/Sidebar/routes/Messages/components/MessageNew';
import PrivateChatSession from 'routes/Sidebar/routes/Messages/components/PrivateChatSession';

import * as API from 'types/api';
import * as UI from 'types/ui';

import {
  SessionProviderDecoratorChildProps,
  SessionProviderDecorator,
} from 'routes/Sidebar/decorators/SessionProviderDecorator';
import IconConstants from 'constants/IconConstants';

import '../style.css';
import { PrivateChatActionMenu } from 'actions/ui/private-chat';

const sessionActions = ['clearChat'];

const Messages: React.FC<SessionProviderDecoratorChildProps<API.PrivateChat>> = (
  props,
) => {
  const { params, sessionT, ...other } = props;
  return (
    <SessionLayout
      activeId={params.id}
      baseUrl="messages"
      newCaption={sessionT.t('new', 'New session')}
      newDescription={sessionT.t('newDesc', 'Open a new private chat session')}
      newIcon={IconConstants.MESSAGES_PLAIN}
      unreadInfoStore={PrivateChatSessionStore}
      editAccess={API.AccessEnum.PRIVATE_CHAT_EDIT}
      uiActions={PrivateChatActionMenu}
      sessionApi={PrivateChatAPIActions as UI.SessionActions<API.PrivateChat>}
      actionIds={sessionActions}
      sessionItemLayout={PrivateChatSession}
      newLayout={MessageNew}
      sessionT={sessionT}
      remoteMenuId="private_chat"
      {...UserItemHandlerDecorator(['browse', 'ignore', 'unignore'])}
      {...other}
    />
  );
};

export default SessionProviderDecorator(
  Messages,
  PrivateChatSessionStore,
  UI.Modules.MESSAGES,
);
