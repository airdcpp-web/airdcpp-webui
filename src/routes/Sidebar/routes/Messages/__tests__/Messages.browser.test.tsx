import { afterEach, beforeAll, beforeEach, describe, expect, test, vi } from 'vitest';
import { RenderResult, waitFor, waitForElementToBeRemoved } from '@testing-library/react';

import { getMockServer } from 'airdcpp-apisocket/tests';

import { renderDataRoutes } from '@/tests/render/test-renderers';

import Messages from '../components/Messages';
import {
  PrivateChat1,
  PrivateChat1MessageMe,
  PrivateChat1MessageOther,
  PrivateChat1MessagesResponse,
  PrivateChat2,
  PrivateChat2MessageOther,
  PrivateChat2MessagesResponse,
} from '@/tests/mocks/api/private-chat';
import { useStoreDataFetch } from '@/components/main/effects/StoreDataFetchEffect';
import PrivateChatConstants from '@/constants/PrivateChatConstants';
import { initCommonDataMocks } from '@/tests/mocks/mock-data-common';
import { setupUserEvent } from '@/tests/helpers/test-form-helpers';

import * as API from '@/types/api';

import SocketNotificationListener from '@/components/main/notifications/SocketNotificationListener';
import {
  NOTIFICATION_EVENT_TYPE,
  NotificationEventEmitter,
} from '@/components/main/notifications/effects/NotificationManager';
import { LocalSettings } from '@/constants/LocalSettingConstants';
import { VIEW_FIXED_HEIGHT } from '@/tests/render/test-containers';

// tslint:disable:no-empty
describe('Private messages', () => {
  let server: ReturnType<typeof getMockServer>;

  /*const addTempShareHandlers = () => {
    server.addRequestHandler(
      'GET',
      ShareConstants.TEMP_SHARES_URL,
      ShareTempItemListResponse,
    );

    const tempItemAdded = server.addSubscriptionHandler(
      ShareConstants.MODULE_URL,
      ShareConstants.TEMP_ITEM_ADDED,
    );

    const tempItemRemoved = server.addSubscriptionHandler(
      ShareConstants.MODULE_URL,
      ShareConstants.TEMP_ITEM_REMOVED,
    );
  }*/

  const getSocket = async () => {
    const commonData = await initCommonDataMocks(server, [
      API.AccessEnum.PRIVATE_CHAT_VIEW,
      API.AccessEnum.PRIVATE_CHAT_EDIT,
      API.AccessEnum.PRIVATE_CHAT_SEND,
    ]);

    // Messages
    const onSession1Read = vi.fn();
    server.addRequestHandler(
      'POST',
      `${PrivateChatConstants.SESSIONS_URL}/${PrivateChat1.id}/messages/read`,
      undefined,
      onSession1Read,
    );

    const onSession2Read = vi.fn();
    server.addRequestHandler(
      'POST',
      `${PrivateChatConstants.SESSIONS_URL}/${PrivateChat2.id}/messages/read`,
      undefined,
      onSession2Read,
    );

    server.addRequestHandler(
      'GET',
      `${PrivateChatConstants.SESSIONS_URL}/${PrivateChat1.id}/messages/0`,
      PrivateChat1MessagesResponse,
    );

    server.addRequestHandler(
      'GET',
      `${PrivateChatConstants.SESSIONS_URL}/${PrivateChat2.id}/messages/0`,
      PrivateChat2MessagesResponse,
    );

    return {
      commonData,
      onSession1Read,
      onSession2Read,
    };
  };

  const renderLayout = async () => {
    const { commonData, ...other } = await getSocket();

    const MessageLayoutTest = () => {
      useStoreDataFetch(true);
      return (
        <>
          <Messages />
          <SocketNotificationListener />
        </>
      );
    };

    const routes = [
      {
        path: '/messages/:session?/:id?/*',
        Component: MessageLayoutTest,
      },
    ];

    const renderData = renderDataRoutes(routes, commonData, {
      routerProps: { initialEntries: ['/messages'] },
      viewType: VIEW_FIXED_HEIGHT,
    });

    return { ...commonData, ...renderData, ...other };
  };

  beforeAll(() => {
    localStorage.setItem('debug', 'true');
  });

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
    localStorage.clear();
  });

  const waitLoaded = async (queryByText: RenderResult['queryByText']) => {
    await waitFor(() => queryByText(/Loading sessions/i));
    await waitForElementToBeRemoved(() => queryByText(/Loading sessions/i));

    await waitFor(() => queryByText(/Loading messages/i));
    await waitForElementToBeRemoved(() => queryByText(/Loading messages/i));
  };

  test('should load messages', async () => {
    const { getByText, queryByText, socket } = await renderLayout();

    await waitLoaded(queryByText);

    // Check content
    await waitFor(() => expect(getByText(PrivateChat1MessageMe.text)).toBeTruthy());

    socket.disconnect();
  });

  test('should handle read messages', async () => {
    const {
      getByText,
      queryByText,
      socket,
      mockStoreListeners,
      onSession1Read,
      onSession2Read,
      sessionStore,
      appStore,
    } = await renderLayout();

    appStore.getState().settings.setValue(LocalSettings.NOTIFY_PM_BOT, true); // The second session is for a bot

    const onNotification = vi.fn();
    NotificationEventEmitter.addEventListener(NOTIFICATION_EVENT_TYPE, onNotification);

    await waitLoaded(queryByText);

    expect(sessionStore.getState().privateChats.activeSessionId).toEqual(PrivateChat1.id);

    // Send message for the active session
    mockStoreListeners.privateChat.chatMessage.fire(
      {
        ...PrivateChat1MessageOther,
        text: 'Chat 1 new message',
        id: PrivateChat1MessageOther.id + 1,
        is_read: false,
      },
      PrivateChat1.id,
    );
    await waitFor(() => expect(onSession1Read).toHaveBeenCalledTimes(1));
    expect(onNotification).toHaveBeenCalledTimes(0);

    // Send message for the background session
    mockStoreListeners.privateChat.chatMessage.fire(
      {
        ...PrivateChat2MessageOther,
        text: 'Chat 2 new message',
        id: PrivateChat2MessageOther.id + 1,
        is_read: false,
      },
      PrivateChat2.id,
    );
    await waitFor(() => expect(onSession2Read).toHaveBeenCalledTimes(0));
    expect(onNotification).toHaveBeenCalledTimes(1);

    // Activate the background session
    const userEvent = setupUserEvent();
    const session2MenuItem = queryByText(PrivateChat2.user.nicks);
    await userEvent.click(session2MenuItem!);

    await waitFor(() => expect(onSession2Read).toHaveBeenCalledTimes(1));
    expect(sessionStore.getState().privateChats.activeSessionId).toEqual(PrivateChat2.id);

    await waitFor(() => expect(getByText(PrivateChat2MessageOther.text)).toBeTruthy());

    socket.disconnect();
  });
});
