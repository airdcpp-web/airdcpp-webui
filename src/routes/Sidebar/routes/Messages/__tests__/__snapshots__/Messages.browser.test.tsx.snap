// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Private messages > highlights > should render magnet 1`] = `
{
  "data": {
    "dupe": {
      "id": "share_full",
      "paths": [
        "/tmp/3825835894",
      ],
    },
    "id": 805,
    "name": "putty test.log",
    "size": 42128,
    "target_name": "putty test.log",
    "tth": "S7WYZE2NGBFNTZQ4J65ZOTSIQJ2GLPJJRVC3IZQ",
    "type": {
      "content_type": "",
      "id": "file",
      "str": "",
    },
    "user": {
      "cid": "HNLUODI2YZ2U7FDWMHFNJU65ERGKUN4MH7GW5LY",
      "description": "[ REG ] ",
      "download_speed": 118750000,
      "email": "",
      "file_count": 55228,
      "flags": [
        "asch",
        "ccpm",
      ],
      "hub_name": "ADC hub 1",
      "hub_session_id": 1,
      "hub_url": "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
      "id": 1213154642,
      "ip4": {
        "country": "",
        "ip": "192.168.50.200",
        "str": "192.168.50.200",
      },
      "ip6": {
        "country": "",
        "ip": "",
        "str": "",
      },
      "nick": "[Hub1]User1",
      "share_size": 7379226364015,
      "supports": [
        "SEGA",
        "ADC0",
        "CCPM",
        "SUD1",
        "TCP4",
        "UDP4",
        "ASCH",
      ],
      "tag": "<AirDC++w 2.13.3,M:A-,H:0/6/1,S:20>",
      "upload_slots": 20,
      "upload_speed": 62500000,
    },
  },
  "method": "POST",
  "path": "queue/bundles/file",
}
`;

exports[`Private messages > highlights > should render releases 1`] = `
{
  "data": {
    "entity_id": "HNLUODI2YZ2U7FDWMHFNJU65ERGKUN4MH7GW5LY",
    "selected_ids": [
      814,
    ],
    "supports": [
      "urls",
      "form",
    ],
  },
  "method": "POST",
  "path": "menus/private_chat_message_highlight/list_grouped",
}
`;
